source/source_integration_test.go

```diff
@@ -0,0 +1,872 @@
+// Copyright © 2022 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package source
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	"cloud.google.com/go/storage"
+	"github.com/conduitio/conduit-connector-google-cloudstorage/config"
+	sourceConfig "github.com/conduitio/conduit-connector-google-cloudstorage/source/config"
+	"github.com/conduitio/conduit-connector-google-cloudstorage/source/position"
+	sdk "github.com/conduitio/conduit-connector-sdk"
+	"github.com/google/uuid"
+	"google.golang.org/api/iterator"
+	"google.golang.org/api/option"
+)
+
+type Object struct {
+	key     string
+	content string
+}
+
+const (
+	projectID = "projectID"
+)
+
+func TestSource_SuccessfulSnapshot(t *testing.T) {
+	client, cfg := prepareIntegrationTest(t)
+
+	ctx := context.Background()
+	testBucket := cfg[config.ConfigKeyGCSBucket]
+	source := &Source{}
+	defer func() {
+		_ = source.Teardown(ctx)
+	}()
+
+	err := source.Configure(ctx, cfg)
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = source.Open(ctx, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	testFiles := addObjectsToTheBucket(ctx, t, testBucket, client, 5)
+
+	// read and assert
+	for _, file := range testFiles {
+		_, err := readAndAssert(ctx, t, source, file)
+		if err != nil {
+			t.Fatalf("unexpected error: %v", err)
+		}
+	}
+
+	_, err = source.Read(ctx)
+	if !errors.Is(err, sdk.ErrBackoffRetry) {
+		t.Fatalf("expected a BackoffRetry error, got: %v", err)
+	}
+}
+
+func TestSource_SnapshotStartFromNonNilPosition(t *testing.T) {
+	client, cfg := prepareIntegrationTest(t)
+
+	ctx := context.Background()
+	testBucket := cfg[config.ConfigKeyGCSBucket]
+	source := &Source{}
+	defer func() {
+		_ = source.Teardown(ctx)
+	}()
+
+	err := source.Configure(ctx, cfg)
+	if err != nil {
+		t.Fatal(err)
+	}
+	p, err := json.Marshal(position.Position{
+		Key:       "file0003",
+		Timestamp: time.Now(),
+		Type:      0,
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+	// set a non nil position
+	err = source.Open(ctx, p)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	testFiles := addObjectsToTheBucket(ctx, t, testBucket, client, 10)
+
+	// read and assert
+	for _, file := range testFiles {
+		// first position is not nil, then snapshot will start from beginning
```

Comment:
If it's not nil, why should it start from beginning?

Reply:
Sure i will elaborate the comment